persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -1738,6 +1751,181 @@ repsertManySql ent n = putManySql' conflictColumns fields ent n
+--
+-- @since 2.12.1.0
+copyField :: PersistField typ => EntityField record typ -> HandleUpdateCollision record
+copyField = CopyField
+
+-- | Postgres specific 'upsertWhere'. This method does the following:
+-- It will insert a record if no matching unique key exists.
+-- If a unique key exists, it will update the relevant field with a user-supplied value, however,
+-- it will only do this update on a user-supplied condition.
+-- For example, here's how this method could be called like such:
+--
+-- upsertWhere record [recordField =. newValue] [recordField /= newValue]
+--
+-- Called thusly, this method will insert a new record (if none exists) OR update a recordField with a new value
+-- assuming the condition in the last block is met.
+-- 
+-- @since 2.12.1.0
+upsertWhere
+  :: ( backend ~ PersistEntityBackend record
+     , PersistEntity record
+     , PersistEntityBackend record ~ SqlBackend
+     , MonadIO m
+     , PersistStore backend
+     , BackendCompatible SqlBackend backend
+     )
+  => record
+  -> [Update record]
+  -> [Filter record]
+  -> ReaderT backend m ()
+upsertWhere record updates filts =
+  upsertManyWhere [record] [] updates filts
+
+-- | Postgres specific 'upsertManyWhere'. This method does the following:
+-- It will insert a record if no matching unique key exists.
+-- If a unique key exists, it will update the relevant field with a user-supplied value, however,
+-- it will only do this update on a user-supplied condition.
+-- For example, here's how this method could be called like such:
+--
+-- upsertManyWhere [record] [recordField =. newValue] [recordField /= newValue]
+--
+-- Called thusly, this method will insert a new record (if none exists) OR update a recordField with a new value
+-- assuming the condition in the last block is met.
+--
+-- -- @since 2.12.1.0
+upsertManyWhere ::
+  forall record backend m.
+  ( backend ~ PersistEntityBackend record,
+    BackendCompatible SqlBackend backend,
+    PersistEntityBackend record ~ SqlBackend,
+    PersistEntity record,
+    MonadIO m
+  ) =>
+  -- | A list of the records you want to insert, or update
+  [record] ->
+  -- | A list of the fields you want to copy over.
+  [HandleUpdateCollision record] ->
+  -- | A list of the updates to apply that aren't dependent on the record being inserted.
+  [Update record] ->
+  -- | A filter condition that dictates the scope of the updates
+  [Filter record] ->
+  ReaderT backend m ()
+upsertManyWhere [] _ _ _ = return ()
+upsertManyWhere records fieldValues updates filters = do
+  conn <- asks projectBackend
+  uncurry rawExecute $
+    mkBulkUpsertQuery records conn fieldValues updates filters
+
+-- | This creates the query for 'upsertManyWhere'. If you
+-- provide an empty list of updates to perform, then it will generate
+-- a dummy/no-op update using the first field of the record. This avoids
+-- duplicate key exceptions.
+mkBulkUpsertQuery
+    :: (PersistEntity record, PersistEntityBackend record ~ SqlBackend)
```

Comment:
I don't think we need the `PersistEntityBackend record ~ SqlBackend` constraint here, either.
```suggestion
    :: (PersistEntity record)
```

Reply:
oh maybe THAT was it